cumulusci/tasks/apex/anon.py

```diff
@@ -1,35 +1,110 @@
 
 
 class AnonymousApexTask(BaseSalesforceApiTask):
-    """ Executes a string of anonymous apex. """
+    """ Executes anonymous apex from a file or string."""
+
+    task_options = {
+        "path": {
+            "description": "The path to an Apex file to run.",
+            "required": False,
+        },
+        "apex": {
+            "description": "A string of Apex to run (after the file, if specified).",
+            "required": False,
+        },
+        "managed": {
+            "description": (
+                "If True, will insert the project's namespace prefix.  "
+                "Defaults to False or no namespace."
+            ),
+            "required": False,
+        },
+        "namespaced": {
+            "description": (
+                "If True, the tokens %%%NAMESPACED_RT%%% and %%%namespaced%%% "
+                "will get replaced with the namespace prefix for Record Types."
+            ),
+            "required": False,
+        },
+    }
+
+    def _validate_options(self):
+        super(AnonymousApexTask, self)._validate_options()
 
-    task_options = {"apex": {"description": "The apex to run.", "required": True}}
+        if not self.options.get('path') and not self.options.get('apex'):
+            raise TaskOptionsError('You must specify either the `path` or `apex` option.')
 
     def _run_task(self):
-        self.logger.info("Executing Anonymous Apex")
+        apex = ''
+        apex_path = self.options.get('path')
+        if apex_path:
+            if not in_directory(apex_path, self.project_config.repo_root):
+                raise TaskOptionsError(
+                    'Please specify a path inside your project repository. '
+                    'You specified: {}'.format(apex_path)
+                )
+            self.logger.info("Executing anonymous Apex from {}".format(apex_path))
+            try:
+                with open(apex_path, 'r') as f:
+                    apex = f.read()
+            except IOError:
+                raise TaskOptionsError("Could not find or read file: {}".format(apex_path))
+        else:
+            self.logger.info('Executing anonymous Apex')
+
+        apex_string = self.options.get('apex')
+        if apex_string:
+            apex = apex + '\n' + apex_string
+
+        apex = self._prepare_apex(apex)
         result = self.tooling._call_salesforce(
             method="GET",
             url="{}executeAnonymous".format(self.tooling.base_url),
-            params={"anonymousBody": self.options["apex"]},
+            params={"anonymousBody": apex},
         )
-        if result.status_code != 200:
-            raise SalesforceGeneralError(url, path, result.status_code, result.content)
+        self._check_result(result)
+
+        self.logger.info("Anonymous Apex Success")
+
+    def _prepare_apex(self, apex):
+        # Replace the email string only if it exists
+        admin_email = self.project_config.dev_config__admin_email
```

Comment:
can we add this convention to the docs somewhere?

Reply:
a new block in `features.rst` called Anonymous Apex that explains this email behavior, as well as the concat behavior and how it is used to call functions of a shared file?

Reply:
I made the task doc generator include extra docs from a `task_docs` attribute on the class, and added a few pointers there. Sorry for putting this in the same commit that includes running black...